controlplane/kubeadm/internal/workload_cluster.go

```diff
@@ -123,37 +153,112 @@ func (w *Workload) ControlPlaneIsHealthy(ctx context.Context) (HealthCheckResult
-			response[name] = err
-			continue
+		var owningMachine *clusterv1.Machine = nil
+		// If there is no owning Machine for the node, setting conditions are no-op.
+		for _, machine := range controlPlane.Machines {
+			if machine.Status.NodeRef == nil {
+				continue
+			}
+			if machine.Status.NodeRef.Name == node.Name {
+				owningMachine = machine
+			}
 		}
 
-		apiServerPodKey := ctrlclient.ObjectKey{
-			Namespace: metav1.NamespaceSystem,
-			Name:      staticPodName("kube-apiserver", name),
+		// If there is any problem with the checks below, set response to any one of the problems.
+		errList := []error{}
+
+		if err := checkNodeNoExecuteTaint(node); err != nil {
+			errList = append(errList, err)
 		}
-		apiServerPod := corev1.Pod{}
-		if err := w.Client.Get(ctx, apiServerPodKey, &apiServerPod); err != nil {
-			response[name] = err
-			continue
+
+		// Check kube-api-server health
+		if _, err = w.reconcilePodStatusCondition(KubeAPIServerPodNamePrefix, node, owningMachine, clusterv1.MachineAPIServerPodHealthyCondition); err != nil {
+			errList = append(errList, err)
 		}
-		response[name] = checkStaticPodReadyCondition(apiServerPod)
 
-		controllerManagerPodKey := ctrlclient.ObjectKey{
-			Namespace: metav1.NamespaceSystem,
-			Name:      staticPodName("kube-controller-manager", name),
+		// Check kube-controller-manager health
+		if _, err = w.reconcilePodStatusCondition(KubeControllerManagerPodNamePrefix, node, owningMachine, clusterv1.MachineControllerManagerHealthyCondition); err != nil {
+			errList = append(errList, err)
 		}
-		controllerManagerPod := corev1.Pod{}
-		if err := w.Client.Get(ctx, controllerManagerPodKey, &controllerManagerPod); err != nil {
-			response[name] = err
+
+		// Check kube-scheduler health
+		if _, err = w.reconcilePodStatusCondition(KubeSchedulerHealthyPodNamePrefix, node, owningMachine, clusterv1.MachineSchedulerPodHealthyCondition); err != nil {
+			errList = append(errList, err)
+		}
+
+		if len(errList) > 0 {
+			response[name] = kerrors.NewAggregate(errList)
 			continue
 		}
-		response[name] = checkStaticPodReadyCondition(controllerManagerPod)
 	}
 
 	return response, nil
 }
 
+type podState struct {
+	provisioning bool
+	unknown      bool
+	ready        bool
+}
+
+// reconcilePodStatusCondition returns error if all podState states are false, i.e., Node is reachable, provisioned, and not in Ready state.
+func (w *Workload) reconcilePodStatusCondition(staticPodPrefix string, node corev1.Node, owningMachine *clusterv1.Machine, conditionType clusterv1.ConditionType) (podState, error) {
+	staticPodKey := ctrlclient.ObjectKey{
+		Namespace: metav1.NamespaceSystem,
+		Name:      staticPodName(staticPodPrefix, node.Name),
+	}
+
+	staticPod := corev1.Pod{}
+	if err := w.Client.Get(context.TODO(), staticPodKey, &staticPod); err != nil {
+		// If there is an error getting the Pod, do not set any conditions.
+		if apierrors.IsNotFound(err) {
+			if owningMachine != nil {
```

Comment:
We don't have a generic action to make in case Machine is nil if we move this check up.

This function is checking Node health even if there is no matching Machine. Do we want to change this behaviour? If there is a Node without a Machine, should we not fail ControlPlaneHealth?